mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+// Copyright 2022 DeepMind Technologies Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mjpc/planners/cem/planner.h"
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <mutex>
+#include <shared_mutex>
+
+#include <absl/random/random.h>
+#include <mujoco/mujoco.h>
+#include "mjpc/array_safety.h"
+#include "mjpc/planners/cem/policy.h"
+#include "mjpc/states/state.h"
+#include "mjpc/trajectory.h"
+#include "mjpc/utilities.h"
+
+namespace mjpc {
+
+namespace mju = ::mujoco::util_mjpc;
+
+// initialize data and settings
+void CEMPlanner::Initialize(mjModel* model, const Task& task) {
+  // delete mjData instances since model might have changed.
+  data_.clear();
+  // allocate one mjData for nominal.
+  ResizeMjData(model, 1);
+
+  // model
+  this->model = model;
+
+  // task
+  this->task = &task;
+
+  // rollout parameters
+  timestep_power = 1.0;
+
+  // sampling noise
+  noise_exploration = GetNumberOrDefault(0.1, model, "sampling_exploration");  // only controls initial variance
+
+  // set number of trajectories to rollout
+  num_trajectory_ = GetNumberOrDefault(10, model, "sampling_trajectories");
+
+  // set number of elite samples
+  n_elites = GetNumberOrDefault(num_trajectory_ / 2, model, "n_elites");  // half rounded down
+
+  if (num_trajectory_ > kMaxTrajectory) {
+    mju_error_i("Too many trajectories, %d is the maximum allowed.",
+                kMaxTrajectory);
+  }
+
+  winner = 0;
+}
+
+// allocate memory
+void CEMPlanner::Allocate() {
+  // initial state
+  int num_state = model->nq + model->nv + model->na;
+
+  // state
+  state.resize(num_state);
+  mocap.resize(7 * model->nmocap);
+  userdata.resize(model->nuserdata);
+
+  // policy
+  int num_max_parameter = model->nu * kMaxTrajectoryHorizon;
+  policy.Allocate(model, *task, kMaxTrajectoryHorizon);
+  previous_policy.Allocate(model, *task, kMaxTrajectoryHorizon);
+
+  // scratch
+  parameters_scratch.resize(num_max_parameter);
+  times_scratch.resize(kMaxTrajectoryHorizon);
+
+  // noise
+  noise.resize(kMaxTrajectory * (model->nu * kMaxTrajectoryHorizon));
+  stdev.resize(kMaxTrajectory * (model->nu * kMaxTrajectoryHorizon));
+  fill(stdev.begin(), stdev.end(), noise_exploration);  // TODO: if this gets called before Initialize, then move this to above
```

Comment:
these two lines initialize the noise standard deviations associated with each control input for all time

Reply:
notably, i'm not fitting non-diagonal terms in the covariance matrix